### PR TITLE
Fix corner cases of go-to-definition & support references for object field (Serokell, Milestone-3)

### PR DESCRIPTION
**Pull request by GoPavel**
This PR contains the following changes:
* Add tests for the `textDocument/definition` request and fix corner cases:
  * occasional jump to import instead of definition
  * jump to the beginning of the declaration instead of defining the symbol name
  * fix handling sequence of qualification `A.B.C.foo`
* Support the `textDocument/references` request for fields of modules and classes

**Note about finding references in the presence of subtyping:**
```motoko
public class Class1() { public func meth() : Nat { 1 }; };
public class Class2() { public func meth() : Int { 2 }; };

private func test(b: Bool): Int {
  let c1                     = if b { Class1() } else { Class2() };
  let c2 : Class2            = Class1();
  let c3 : {meth Nat -> Int} = Class1(); 
  return c1.meth() + c2.meth() + c3.meth();
}
```
The code about is showing that finding all references of a symbol  (i.e. `meth` ) absolutely precise is impossible. Thus, we made a reasonable assumption: if the problem contains any subtyping cast, then the reference analysis is over-approximating: any reference `Class1.meth` is also counted as a reference `Class2.meth` as well. Also, go-to-definition will show two definitions.

Related PRs:
* `motoko`: https://github.com/dfinity/motoko/pull/5076
* `node-motoko`: https://github.com/dfinity/node-motoko/pull/128

### CODE REVIEW

**Comment by rvanasa** [comment]
I opened a ticket about the CI error to see if we can get that resolved for when https://github.com/dfinity/motoko/pull/5076 is ready to merge.

File: src/server/handlers.ts [hunk 1]
```diff
@@ -1484,7 +1505,7 @@ export const addHandlers = (connection: Connection, redirectConsole = true) => {
                     } else if (!isSameLine) {
                         docs.push(codeSnippet(source));
                     }
-                    docs.push(doc);
+                    docs = docs.concat(nodeDocs);
```
[the remark this answers is not in the document]
Comment (heitor-lassarote) [reply src/server/handlers.ts]: Done.

**Comment by heitor-lassarote** [comment]
Force-pushed to rebase with `master`.

File: src/server/navigation.ts [hunk 2]
```diff
@@ -149,38 +196,66 @@ export function findDefinition(
...
         console.log(
             'Definition not found for reference path:',
-            path,
+            relatedPath,
             `(${node.name})`,
         );
         return;
     }
-    return definition;
+    return Array.isArray(definition) ? definition : [definition];
```
[the remark this answers is not in the document]
Comment (heitor-lassarote) [reply src/server/navigation.ts]: Done.

File: src/server/navigation.ts [hunk 3]
```diff
@@ -291,14 +374,85 @@ export function followImport(
...
+    }
+    return searchDotE(reference.node);
+}
+
 function search(
     context: Context,
     reference: Reference,
     path: Search[],
-): Definition | undefined {
+): Definition | Definition[] | undefined {
```
[the remark this answers is not in the document]
Comment (heitor-lassarote) [reply src/server/navigation.ts]: Changed to just `Definition[]`. With the current implementation, it's possible we'd get `undefined` even with valid input, in case there is a bug in definitions or something missing, so I preferred just `Definition[]`.

File: src/server/navigation.ts [hunk 4]
```diff
@@ -51,8 +90,10 @@ export function findMostSpecificNodeForPosition(
         if (
             !node ||
             scoreFn(n) > scoreFn(node) ||
-            nLines < nodeLines ||
-            (nLines == nodeLines && nChars < nodeChars)
+            (scoreFn(n) === scoreFn(node) && nLines < nodeLines) ||
+            (scoreFn(n) === scoreFn(node) &&
```
[the remark this answers is not in the document]
Comment (heitor-lassarote) [reply src/server/navigation.ts]: Done. We don't need to call `scoreFn(node)`, if under the assumption that `scoreFn` is pure/deterministic.

File: src/server/syntax.spec.ts [hunk 5]
```diff
@@ -120,3 +130,106 @@ describe('incorrect syntax', () => {
         expectWithFields(prog.exportFields[0].exp, ['x']);
     });
 });
+
+/*
+// TODO: Uncomment those tests when _every_ AST node will be annotated with a
+// `typeRep !== undefined`.
```
[the remark this answers is not in the document]
Comment (heitor-lassarote) [reply src/server/syntax.spec.ts:136]: In theory, all that needs to be done is to serialize every node in Motoko. This is what we were doing at first, but @crusso pointed out that there is no sharing between types, which could potentially increase memory and processor usage, which turned out to be true according to benchmarks (hence, now we only do it for the LHS of `DotE` nodes, which is just what the extension uses).

So in practice, the steps should be to develop a sharing mechanism to avoid duplication (perhaps by annotating every type with some kind of ID, like a hash), and during serialization, consider, for example, exporting a table of types.

The design of this is up for debate, one way might be to change `Mo_types.Type.t` to be a record `{ it : typ; note : hash }` for example, although this seems very intrusive, as the use of `t` through the compiler is very extensive. We could then have a new AST that mimics `Type.t`, but replacing every recursive usage of `Type.t` with `hash`, and put everything inside a `Hashtbl.t` to be exported together with the AST (whether we'd have to create a new datatype or this will would just exist during serialization is up for more thinking).

With this idea, something like `(Tup (Tup Int Nat) (Tup Int Nat) Text)` would produce a table `((0 Int) (1 Nat) (2 (Tup 0 1)) (3 Text) (4 (Tup 2 2 3)))` and we'd see these numbers in the AST to indicate the types.

However, since we're sending the AST to JS as an S-expression (and not as a huge `string`), I think there is a simpler idea that could work: we still store IDs for the types, and during serialization, we instead build a table relating these IDs to its serialized S-expression, and every time we see a `Type.t`, we check if it's in the table or not and fetch/serialize appropriately. This would also avoid having to export a table (and having a new types AST if we find that to be necessary). The first idea might be better if we need to send it as JSON or some other format that doesn't support sharing, which is not our case.

But all of this is just from the top of my head, all of this is up for debate, and more thinking. For implementing requests such as "Go to type definition", it might be pretty useful to come back to it, though.

By the way, currently, Motoko serializes `type` for every node, which introduces a `string` with the pretty-printed type. The above solution could also be used to resolve this and avoid excessive calls to the pretty-printer.

File: src/server/navigation.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (rvanasa) [reply src/server/navigation.ts]: Thanks! Good point about only needing to call `scoreFn(node)` once.

File: src/server/syntax.spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (rvanasa) [reply src/server/syntax.spec.ts:136]: Thanks for the explanation! I like the approach of marshalling using type IDs. It seems reasonable to revisit this in the future, such as when implementing go-to-type-definition as you mentioned. 

> By the way, currently, Motoko serializes type for every node, which introduces a string with the pretty-printed type. The above solution could also be used to resolve this and avoid excessive calls to the pretty-printer.

That is a good point; +1 to this idea.

**Comment by heitor-lassarote** [comment]
Rebased with the latest master and added your changes. Hope all is fine to merge this PR too. :slightly_smiling_face:

**Comment by heitor-lassarote** [comment]
Seems like we should increase the timeout for tests, I've set them to 60 seconds.

**Review by rvanasa — APPROVED** [review]
